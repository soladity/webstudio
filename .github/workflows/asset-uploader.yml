name: Asset Uploader
on:
  push:
    paths:
      - "packages/asset-uploader/**/*"
      - ".github/workflows/asset-uploader.yml"
      - ".github/actions/publish/action.yaml"
    branches:    
      - main
jobs:
  publish:
    runs-on: ubuntu-latest
    steps:
      - uses: actions/checkout@v3
      - uses: ./.github/actions/publish
        with:
          workdir: ./packages/asset-uploader
          filter: "@webstudio-is/asset-uploader"
          token: ${{ secrets.NPM_TOKEN }}
