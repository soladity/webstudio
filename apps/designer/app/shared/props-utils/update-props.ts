import {
  type AllUserProps,
  type UserPropsUpdates,
} from "@webstudio-is/react-sdk";

export const updateAllUserPropsMutable = (
  allUserProps: AllUserProps,
  { instanceId, propsId, treeId, updates }: UserPropsUpdates
) => {
  if (instanceId in allUserProps === false) {
    allUserProps[instanceId] = {
      id: propsId,
      instanceId,
      treeId,
      props: [],
    };
  }
  const instanceProps = allUserProps[instanceId];
  for (const update of updates) {
    const propIndex = instanceProps.props.findIndex(
      ({ id }) => id === update.id
    );

    if (propIndex === -1) {
      instanceProps.props.push(update);
    } else {
      const prop = instanceProps.props[propIndex];

      prop.prop = update.prop;
      prop.type = update.type;
      prop.value = update.value;
      prop.required = update.required;
    }
  }
};
