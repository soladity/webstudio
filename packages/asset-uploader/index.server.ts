export * from "./src/db";
export * from "./src/upload";
export * from "./src/delete";
export * from "./src/utils/format-asset";
