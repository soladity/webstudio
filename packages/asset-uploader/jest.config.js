// eslint-disable-next-line
const base = require("jest-config/jest.config.js");

module.exports = {
  ...base,
};
