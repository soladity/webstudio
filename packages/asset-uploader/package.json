{
  "name": "@webstudio-is/asset-uploader",
  "version": "0.75.0",
  "description": "Webstudio Asset Uploader",
  "author": "Webstudio <github@webstudio.is>",
  "homepage": "https://webstudio.is",
  "type": "module",
  "scripts": {
    "typecheck": "tsc --noEmit --emitDeclarationOnly false",
    "test": "NODE_OPTIONS=--experimental-vm-modules jest",
    "checks": "pnpm typecheck && pnpm lint && pnpm test",
    "dev": "build-package --watch",
    "build": "build-package",
    "dts": "tsc --project tsconfig.dts.json",
    "lint": "eslint ./src --ext .ts,.tsx --max-warnings 0"
  },
  "dependencies": {
    "@aws-sdk/client-s3": "^3.131.0",
    "@aws-sdk/lib-storage": "^3.131.0",
    "@webstudio-is/fonts": "workspace:^",
    "@webstudio-is/prisma-client": "workspace:^",
    "@webstudio-is/trpc-interface": "workspace:^",
    "immer": "^9.0.12",
    "nanoid": "^3.3.6",
    "sharp": "^0.30.7",
    "warn-once": "^0.1.1"
  },
  "devDependencies": {
    "@jest/globals": "^29.3.1",
    "@types/node": "^18.11.18",
    "@types/sharp": "^0.30.4",
    "@webstudio-is/jest-config": "workspace:^",
    "@webstudio-is/scripts": "workspace:^",
    "@webstudio-is/tsconfig": "workspace:^",
    "jest": "^29.3.1",
    "typescript": "5.1.3",
    "zod": "^3.21.4"
  },
  "exports": {
    ".": {
      "source": "./src/index.ts",
      "types": "./lib/types/index.d.ts",
      "import": "./lib/index.js",
      "require": "./lib/cjs/index.js"
    },
    "./index.server": {
      "source": "./src/index.server.ts",
      "types": "./lib/types/index.server.d.ts",
      "import": "./lib/index.server.js"
    }
  },
  "files": [
    "lib/*",
    "src/*",
    "!*.test.*"
  ],
  "license": "MIT",
  "private": false,
  "sideEffects": false
}
