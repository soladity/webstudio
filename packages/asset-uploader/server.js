module.exports = require("./lib/index.server");
