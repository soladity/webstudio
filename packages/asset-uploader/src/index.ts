import path from "path";
import {
  unstable_createFileUploadHandler,
  unstable_parseMultipartFormData,
} from "@remix-run/node";
import { s3UploadHandler } from "./targets/s3/handler";
import { uploadToS3 } from "./targets/s3/uploader";
import { uploadToDisk } from "./targets/disk/upload";
import { assetEnvVariables, fsEnvVariables, s3EnvVariables } from "./schema";

const isS3Upload = s3EnvVariables.safeParse(process.env).success;
const fsUploadVars = fsEnvVariables.parse(process.env);
const commonUploadVars = assetEnvVariables.parse(process.env);

// user inputs the max value in mb and we transform it to bytes
export const MAX_UPLOAD_SIZE = parseInt(commonUploadVars.MAX_UPLOAD_SIZE) * 1e6;

export const uploadAssets = async ({
  request,
  projectId,
  db,
  dirname,
}: {
  request: Request;
  projectId: string;
  db: unknown;
  dirname: string;
}) => {
  const uploads = path.join(dirname, "../public");
  const directory = path.join(uploads, fsUploadVars.FILE_UPLOAD_PATH);
  const formData = await unstable_parseMultipartFormData(
    request,
    isS3Upload
      ? (file) =>
          s3UploadHandler({
            file,
            maxPartSize: MAX_UPLOAD_SIZE,
          })
      : unstable_createFileUploadHandler({
          maxPartSize: MAX_UPLOAD_SIZE,
          directory,
          file: ({ filename }) => filename,
        })
  );
  if (isS3Upload) {
    return await uploadToS3({
      projectId,
      formData,
      db,
    });
  } else {
    return await uploadToDisk({
      projectId,
      formData,
      folderInPublic: fsUploadVars.FILE_UPLOAD_PATH,
      db,
    });
  }
};
