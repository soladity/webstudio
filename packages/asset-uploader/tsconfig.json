{
  "extends": "../../node_modules/@webstudio-is/tsconfig/base.json",
  "include": ["./src/**/*.ts"],
  "compilerOptions": {
    "module": "commonjs",
    "target": "ES2020",
    "strict": true,
    "baseUrl": ".",
    "outDir": "./lib",
    "rootDir": "./src"
  },
  "references": [{ "path": "../fonts" }, { "path": "../prisma-client" }]
}
