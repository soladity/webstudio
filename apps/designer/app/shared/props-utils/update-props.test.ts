import { Asset } from "@webstudio-is/asset-uploader";
import {
  type AllUserProps,
  type UserPropsUpdates,
} from "@webstudio-is/react-sdk";
import { updateAllUserPropsMutable } from "./update-props";

describe("Update props", () => {
  test("with empty updates", () => {
    const propsMap: AllUserProps = {
      instanceId: {
        id: "id",
        instanceId: "instanceId",
        treeId: "treeId",
        props: [],
      },
    };
    const update: UserPropsUpdates = {
      propsId: "id",
      instanceId: "instanceId",
      treeId: "treeId",
      updates: [],
    };
    updateAllUserPropsMutable(propsMap, update);
    expect(propsMap).toMatchSnapshot();
  });

  test("with added prop", () => {
    const propsMap: AllUserProps = {
      instanceId: {
        id: "id",
        instanceId: "instanceId",
        treeId: "treeId",
        props: [],
      },
    };
    const update: UserPropsUpdates = {
      propsId: "id",
      instanceId: "instanceId",
      treeId: "treeId",
      updates: [{ id: "propId", prop: "a", value: "1" }],
    };
    updateAllUserPropsMutable(propsMap, update);
    expect(propsMap).toMatchSnapshot();
  });

  test("with updated prop value", () => {
    const propsMap: AllUserProps = {
      instanceId: {
        id: "id",
        instanceId: "instanceId",
        treeId: "treeId",
        props: [{ id: "propId", prop: "a", value: "1" }],
      },
    };
    const update: UserPropsUpdates = {
      propsId: "id",
      instanceId: "instanceId",
      treeId: "treeId",
      updates: [{ id: "propId", prop: "a", value: "2" }],
    };
    updateAllUserPropsMutable(propsMap, update);
    expect(propsMap).toMatchSnapshot();
  });

  test("with renamed prop name", () => {
    const propsMap: AllUserProps = {
      instanceId: {
        id: "id",
        instanceId: "instanceId",
        treeId: "treeId",
        props: [{ id: "propId", prop: "a", value: "1" }],
      },
    };
    const update: UserPropsUpdates = {
      propsId: "id",
      instanceId: "instanceId",
      treeId: "treeId",
      updates: [{ id: "propId", prop: "b", value: "1" }],
    };
    updateAllUserPropsMutable(propsMap, update);
    expect(propsMap).toMatchSnapshot();
  });

  test("with changed value to asset type", () => {
    const propsMap: AllUserProps = {
      instanceId: {
        id: "id",
        instanceId: "instanceId",
        treeId: "treeId",
        props: [{ id: "propId", prop: "a", value: "1" }],
      },
    };
    const update: UserPropsUpdates = {
      propsId: "id",
      instanceId: "instanceId",
      treeId: "treeId",
      updates: [
        {
          id: "propId",
          prop: "a",
          value: "test",
          asset: { id: "assetId" } as Asset,
        },
      ],
    };
    updateAllUserPropsMutable(propsMap, update);
    expect(propsMap).toMatchInlineSnapshot(`
      {
        "instanceId": {
          "id": "id",
          "instanceId": "instanceId",
          "props": [
            {
              "asset": {
                "id": "assetId",
              },
              "id": "propId",
              "prop": "a",
              "value": "test",
            },
          ],
          "treeId": "treeId",
        },
      }
    `);
  });
});
