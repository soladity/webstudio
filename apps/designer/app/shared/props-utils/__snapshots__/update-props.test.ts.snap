// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Update props with added prop 1`] = `
{
  "instanceId": {
    "id": "id",
    "instanceId": "instanceId",
    "props": [
      {
        "id": "propId",
        "prop": "a",
        "value": "1",
      },
    ],
    "treeId": "treeId",
  },
}
`;

exports[`Update props with empty updates 1`] = `
{
  "instanceId": {
    "id": "id",
    "instanceId": "instanceId",
    "props": [],
    "treeId": "treeId",
  },
}
`;

exports[`Update props with renamed prop name 1`] = `
{
  "instanceId": {
    "id": "id",
    "instanceId": "instanceId",
    "props": [
      {
        "id": "propId",
        "prop": "b",
        "value": "1",
      },
    ],
    "treeId": "treeId",
  },
}
`;

exports[`Update props with updated prop value 1`] = `
{
  "instanceId": {
    "id": "id",
    "instanceId": "instanceId",
    "props": [
      {
        "id": "propId",
        "prop": "a",
        "value": "2",
      },
    ],
    "treeId": "treeId",
  },
}
`;
